OSDOCS-18945: adds troubleshooting MCP gateway (#110076)
openshift/openshift-docs: modules/proc-mcp-gateway-ts-requests-fail-or-bypass-router.adoc

// Module included in the following assemblies:
//
// *mcp_gateway_config/mcp-gateway-troubleshooting.adoc

:_mod-docs-content-type: PROCEDURE
[id="proc-mcp-gateway-ts-requests-fail-or-bypass-router_{context}"]
= Troubleshooting requests failing or bypassing the router

[role="_abstract"]
When you are certain that an `MCPGatewayExtension` custom resource (CR) exists for your MCP server, but requests either fail or bypass the router, it might mean that the `EnvoyFilter` CR is not applied properly. You can take several steps to troubleshoot the problem.

> The `EnvoyFilter` CR is automatically created in the `Gateway` CR's namespace by the MCP gateway controller component when an `MCPGatewayExtension` CR is `Ready`.

Reviewer: Is there an attribute for "MCP gateway controller"?

Author: No

[IMPORTANT]
====

> You can look closely at the `EnvoyFilter` during deep troubleshooting, but do not manually edit or delete the CR.

Reviewer: ```suggestion
You can look closely at the `EnvoyFilter` CR during deep troubleshooting, but do not manually edit or delete the CR.
```

====

.Prerequisites

* You installed {mcpg}.
* You installed the {oc-first}.
* You configured a `Gateway` object.
* You configured an `HTTPRoute` object for the gateway.
* You registered an MCP server.

.Procedure

> . Ensure that the `Gateway` object exists and is in the expected namespace by checking the general `Gateway` object configuration by using the following command:

Reviewer: ```suggestion
. Ensure that the `Gateway` object exists and is in the expected namespace by checking the general `Gateway` object configuration by running the following command:
```
Looks like you typically use "by running the following command".

+
[source,terminal]
----
$ oc get gateway -A
----
+

> This command returns general information about all `Gateway` objects in the cluster. If the `Gateway` object you are troubleshooting does exist, the command returns the `gatewayClassName` is it using, whether or not it has an IP address or hostname assigned, and a `status`, such as `Ready`, `Programmed`, or `Pending`.

Reviewer: ```suggestion
This command returns general information about all `Gateway` objects in the cluster. If the `Gateway` object you are troubleshooting exists, the command returns the `gatewayClassName` is it using, whether or not it has an IP address or hostname assigned, and a `status`, such as `Ready`, `Programmed`, or `Pending`.
```

. Check that the `EnvoyFilter` exists in the `Gateway` namespace by running the following command:
+
[source,terminal,subs="+quotes"]
----
$ oc get envoyfilter -n _<gateway_namespace>_ -l app.kubernetes.io/managed-by=mcp-gateway-controller
----
+
* Replace `_<gateway_namespace>_` with the namespace where the `Gateway` object is applied.

. Verify the `EnvoyFilter` configuration by running the following command:
+
[source,terminal,subs="+quotes"]
----
$ oc describe envoyfilter -n _<gateway_namespace>_ -l app.kubernetes.io/managed-by=mcp-gateway-controller
----
+
* Replace `_<gateway_namespace>_` with the namespace where the `Gateway` object is applied.
* The `workloadSelector` labels must match your `Gateway` pods, or your policies are bypassed.

. Compare the `EnvoyFilter` labels against your pod labels by running the following command:
+
[source,terminal,subs="+quotes"]
----
$ oc get pods -n _<gateway_namespace>_ --show-labels
----
+
* Replace `_<gateway_namespace>_` with the namespace where the `Gateway` object is applied.

. Identify the port that the `Gateway` object is configured to use by running the following command:
+
[source,terminal,subs="+quotes"]
----

> $ oc get gateway <gateway_name> -n _<gateway_namespace>_ -o jsonpath='{range .spec.listeners[*]}{.name}{": "}{.port}{"\n"}{end}'

Reviewer: ```suggestion
$ oc get gateway _<gateway_name>_ -n _<gateway_namespace>_ -o jsonpath='{range .spec.listeners[*]}{.name}{": "}{.port}{"\n"}{end}'
```

----
+
* Replace `_<gateway_name>_` with the name of the `Gateway` object.
* Replace `_<gateway_namespace>_` with the namespace where the `Gateway` object is applied.

. Verify the `EnvoyFilter` chain binding by running the following command:
+
[source,terminal,subs="+quotes"]
----

> $ oc exec -n _<gateway_namespace>_ deploy/_<gateway_name>_-istio -- curl -s localhost:15000/config_dump | jq '.configs[] | select(.["@type"] | contains("ListenersConfigDump")) | .dynamic_listeners[] | select(.name | contains(":_<gateway_port>_"))'

Reviewer 2: That ":_<gateway_port>_" piece, the colon (':') should not be there, the underscore ('_') should be there instead. The name of the dynamic listener is like '0.0.0.0_8080' (colon is used in URLs)

> $ oc exec -n _<gateway_namespace>_ deploy/_<gateway_name>_-istio -- curl -s localhost:15000/config_dump | jq '.configs[] | select(.["@type"] | contains("ListenersConfigDump")) | .dynamic_listeners[0.0.0.0_8080] | select(.name | contains("_<gateway_port>_"))'

Reviewer 2: The commands like this one does not work because `curl` binary is missing but even if there's no time to fix them let's keep them anyway - they still show valuable info on how to debug issues

Author: should we do something different, like:
`$ oc debug -n <gateway_namespace> deploy/<gateway_name>-istio -- curl -s localhost:15000/config_dump`

Reviewer 2: @ShaunaDiaz Unfortunately that won't work. What would work:

`oc port-forward deploy/_<gateway_deployment_name>_ -n _<gateway_system>_ 15000:15000`

- where gateway_deployment_name would be the name of Deployment related to Gateway object. Typically it is either `<gateway-name>-istio` or `<gateway-name>-openshift-default` depending on whether full OSSMv3 was installed or just OpenShift's Ingress operator was used.

Then one can access localhost:15000 from a different terminal window:
`curl -s localhost:15000/config_dump...`

// Module included in the following assemblies:
//
// *mcp_gateway_config/mcp-gateway-troubleshooting.adoc

:_mod-docs-content-type: PROCEDURE
[id="proc-mcp-gateway-ts-requests-fail-or-bypass-router_{context}"]
= Troubleshooting requests failing or bypassing the router

[role="_abstract"]
When you are certain that an `MCPGatewayExtension` custom resource (CR) exists for your MCP server, but requests either fail or bypass the router, it might mean that the `EnvoyFilter` CR is not applied properly. You can take several steps to troubleshoot the problem.

The `EnvoyFilter` CR is automatically created in the `Gateway` CR's namespace by the MCP gateway controller component when an `MCPGatewayExtension` CR is `Ready`.

[IMPORTANT]
====
You can look closely at the `EnvoyFilter` during deep troubleshooting, but do not manually edit or delete the CR.
====

.Prerequisites

* You installed {mcpg}.

> * You installed {prodname}.

Reviewer 2: RHCL not needed

* You configured a `Gateway` object.
* You configured an `HTTPRoute` object for the gateway.
* You installed the {oc-first}.
* You registered an MCP server.

.Procedure

. Ensure that the `Gateway` object exists and is in the expected namespace by checking the general `Gateway` object configuration by using the following command:
+
[source,terminal]
----
$ oc get gateway -A
----
+
This command returns general information about all `Gateway` objects in the cluster. If the `Gateway` object you are troubleshooting does exist, the command returns the `gatewayClassName` is it using, whether or not it has an IP address or hostname assigned, and a `status`, such as `Ready`, `Programmed`, or `Pending`.

. Verify that the `MCPGatewayExtension` CR is `Ready` by running the following command:
+
[source,terminal]
----

> $ oc get mcpgatewayextension -A`

Reviewer 2: remove trailing backtick

----

. Verify that a `ReferenceGrant` CR exists if the `MCPGatewayExtension` CR is in a different namespace than the `Gateway` object by running the following command:
+
[source,terminal,subs="+quotes"]
----

> $ oc get referencegrant _<referencegrant_name>_ -n _<referencegrant_namespace>_ -o yaml

Reviewer 2: ReferenceGrant needs to be in the namespace where Gateway object is applied so better to have `-n _<gateway_system>_` here imo.

+
[source,terminal,subs="+quotes"]
----
$ oc describe envoyfilter -n _<gateway_namespace>_ -l app.kubernetes.io/managed-by=mcp-gateway-controller
----
+
* Replace `_<gateway_namespace>_` with the namespace where the `Gateway` object is applied.
* The `workloadSelector` labels must match your `Gateway` pods, or your policies are bypassed.

. Compare the `EnvoyFilter` labels against your pod labels by running the following command:
+
[source,terminal,subs="+quotes"]
----
$ oc get pods -n _<gateway_namespace>_ --show-labels
----
+
* Replace `_<gateway_namespace>_` with the namespace where the `Gateway` object is applied.

. Identify the port that the `Gateway` object is configured to use by running the following command:
+
[source,terminal,subs="+quotes"]
----
$ oc get gateway <gateway_name> -n _<gateway_namespace>_ -o jsonpath='{range .spec.listeners[*]}{.name}{": "}{.port}{"\n"}{end}'
----
+
* Replace `_<gateway_name>_` with the name of the `Gateway` object.
* Replace `_<gateway_namespace>_` with the namespace where the `Gateway` object is applied.

. Confirm that Envoy opened the `Gateway`-specified port by running the following command:
+
[source,terminal,subs="+quotes"]
----

> $ oc exec -n <gateway_namespace> deploy/<gateway_name>-istio -- curl -s localhost:15000/listeners | grep -E ":[0-9]+"

Reviewer 2: Again, that `-istio` piece is only valid if full OSSMv3 is used. If OpenShift's Ingress controller is used then it is `-openshift-default`. Another issue is that on my OpenShift this fails with `executable file `curl` not found in $PATH: No such file or directory`

To be honest I have no idea how to fix this command and the other similar commands a bit below.

Author: I can comment them out and we can come back if we find something that works.

Reviewer 2: @Patryk-Stefanski @jasonmadigan any idea how to change these commands so that they actually work?

One possible solution would be to use port forwarding:
`oc port-forward deploy/_<gateway_deployment_name>_ -n _<gateway_system>_ 15000:15000`

- where `gateway_deployment_name` would be either `mcp-gateway-istio` or `mcp-gateway-openshift-default` depending on whether full OSSMv3 was installed or just OpenShift's Ingress operator was used

Then one can access localhost:15000 from a local machine via curl:
`curl -s localhost:15000/config_dump...`

Wdyt?

----

. Verify the `EnvoyFilter` chain binding by running the following command:
+
[source,terminal,subs="+quotes"]
----

> $ oc exec -n _<gateway_namespace>_ deploy/_<gateway_name>_-istio -- curl -s localhost:15000/config_dump | jq '.configs[] | select(.["@type"] | contains("ListenersConfigDump")) | .dynamic_listeners[] | select(.name | contains(":<PORT_FROM_STEP_6>"))'

Reviewer 2: Not sure what step 6 it is meant here. It is probably step 9 now

----
+
* Replace `_<gateway_name>_` with the name of the `Gateway` object.
* Replace `_<gateway_namespace>_` with the namespace where the `Gateway` object is applied.

> * Check for `envoy.filters.http.ext_authz` for Authorino or `envoy.filters.http.ratelimit` for Limitador within the specific port config.

Reviewer 2: These seems to be related to AuthPolicy and RateLimitPolicy rather than general work of EnvoyFilter. I think users should look at `envoy.filters.http.ext_proc`

. Check the Istio gateway pod configuration by running the following command:
+
[source,terminal,subs="+quotes"]
----

> $ oc exec -n _<gateway_namespace>_ deploy/_<gateway_name>_-istio -- curl localhost:15000/config_dump | grep ext_proc

Reviewer 2: Due to that port-forwarding done earlier this is just:
`curl -s localhost:15000/config_dump | grep ext_proc`

No need for that `oc exec...` part at all.

----
+
* Replace `_<gateway_namespace>_` with the namespace where the `Gateway` object is applied.
* Replace `_<gateway_name>_` with the names of the `Gateway` object.
* If this command returns empty, the `EnvoyFilter` CR is not active on this pod. Traffic is bypassing your policies.

. Restart the Istio gateway to force a configuration reload by running the following command:
+
[source,terminal,subs="+quotes"]
----

> $ oc rollout restart deployment/_<gateway_name>_-istio -n _<gateway_namespace>_`

Reviewer 2: Remove trailing backticks
